text/0000-Component-Manager-Bounds.md

```diff
@@ -0,0 +1,177 @@
+
+### `elementHook` capability
+
+The primary way of introducing new functionality to the component manager is through capabilities. To allow custom component managers to announce that they support element related hooks, we will introduce an `elementHook` flag to the capabilities. In doing so, bounds will be dispatched to a `didRenderLayout` method on the manager and `willDestroyLayout` will be called during destruction.
+
+```js
+import { capabilities } from '@ember/component';
+import EmberObject from '@ember/object';
+
+export default EmberObject.extend({
+  capabilities: capabilities('3.5', {
+    elementHook: true
+  }),
+
+  // ...
+});
+```
+
+### `didRenderLayout` hook
+Once the rendering engine has constructed the bounds for the component the `didRenderLayout` hook will be called with the `component` and a `Bounds` instance.
+
+
+```js
+import { capabilities } from '@ember/component';
+import EmberObject from '@ember/object';
+
+export default EmberObject.extend({
+  capabilities: capabilities('3.5', {
+    elementHook: true
+  }),
+
+  didRenderLayout(component, bounds) {
+    component.bounds = bounds;
+    component.didInsertElement();
+  }
+  // ...
+});
+```
+
+#### Timing Semantics
+
+`didRenderLayout` has the following timing semantics:
+
+**Always**
+- called **before** `didCreateComponent`
+- called **after** the children's `didRenderLayout` hook is called
+- called **after** DOM insertion
+
+**May or May Not**
+- be called in the same tick as DOM insertion
+- be called in the same tick as `didCreateComponent`
+- have the the parent's children fully initialized in DOM
+
+In other words, components **should not** rely on the timing of DOM insertion outside of the component's own bounds.
+
+### `willDestroyLayout` hook
+
+To balance `didRenderLayout` we will expose a hook called `willDestroyLayout` that is invoked during destruction. It will receive the `component` instance.
+
+```js
+import { capabilities } from '@ember/component';
+import EmberObject from '@ember/object';
+
+export default EmberObject.extend({
+  capabilities: capabilities('3.5', {
+    elementHook: true
+  }),
+
+  didRenderLayout(component, bounds) {
+    component.bounds = bounds;
+    component.didInsertElement();
+  },
+
+  willDestroyLayout(component) {
+    component.willDestroyElement();
+  }
+  // ...
+});
+```
+
+#### Timing Semantics
+
+`willDestroyLayout` has the following timing semantics:
+
+**Always**
+- called **before** `destroyComponent`
+- called **after** the children's `willDestroyLayout` hook is called
+
+**May or May Not**
+- be called in the same tick as DOM removal
+- be called in the same tick as `destroyComponent`
+
+### `Bounds` instance
+
+The `Bounds` instance that is passed to `didRenderLayout` has the following interface.
+
+```ts
+
+type FirstNode = Node;
+type LastNode = Node;
+
+interface Bounds {
```

Comment:
`Bounds` doesn't get `.parentElement`?

The interface in glimmer-vm does have `parentElement` property:

https://github.com/glimmerjs/glimmer-vm/blob/e30b3d64232a0972365c8cbb1dd8e236a0f33dcf/packages/%40glimmer/runtime/lib/bounds.ts#L4-L9

Reply:
Due to the lack of timing guarantees of construction of the parent do think this actually useful?

Reply:
I think we need to provide access to the parent. What we cannot guarantee is that the parent's children is fully constructed. (And possibly that the grandparent may not be construed yet.)

By the way, I think you meant to say:

```ts
interface Bounds {
  firstNode: Node | Bounds;
  lastNode: Node | Bounds;
 }
```